state/protocol/badger/mutator.go

```diff
@@ -871,31 +879,36 @@ func (m *FollowerState) MarkValid(blockID flow.Identifier) error {
 		return fmt.Errorf("can only mark block as valid whose parent is valid")
 	}
 
-	err = operation.RetryOnConflict(
-		m.db.Update,
-		operation.SkipDuplicates(
-			operation.InsertBlockValidity(blockID, true),
-		),
-	)
-	if err != nil {
-		return fmt.Errorf("could not mark block as valid (%x): %w", blockID, err)
-	}
-
-	// root blocks and blocks below the root block are considered as "processed",
-	// so we don't want to trigger `BlockProcessable` event for them.
 	parent, err := m.headers.ByBlockID(parentID)
 	if err != nil {
 		return fmt.Errorf("could not retrieve block header for %x: %w", parentID, err)
 	}
-	var rootHeight uint64
-	err = m.db.View(operation.RetrieveRootHeight(&rootHeight))
-	if err != nil {
-		return fmt.Errorf("could not retrieve root block's height: %w", err)
-	}
-	if rootHeight >= parent.Height {
+
+	err = operation.RetryOnConflict(m.db.Update, func(tx *badger.Txn) error {
+		// insert block validity for this block
+		err = operation.SkipDuplicates(operation.InsertBlockValidity(blockID, true))(tx)
+		if err != nil {
+			return fmt.Errorf("could not insert validity for block (id=%x, height=%d): %w", blockID, header.Height, err)
+		}
+
+		// root blocks and blocks below the root block are considered as "processed",
+		// so we don't want to trigger `BlockProcessable` event for them.
+		var rootHeight uint64
+		err = m.db.View(operation.RetrieveRootHeight(&rootHeight))
+		if err != nil {
+			return fmt.Errorf("could not retrieve root block's height: %w", err)
+		}
+		// trigger BlockProcessable for parent blocks above root height
```

Comment:
```suggestion
		// a root block is considered as processed. Since we are triggering BlockProcessable for 
		// parent blocks, it should not be triggered if parent block is a root block. Therefore, 
		// trigger BlockProcessable for parent blocks above root height
```